### PR TITLE
add test for main thread thread-local destructors

### LINKED ISSUE #28129: TLS destructors on the main thread are a little sketchy
There are some platforms where TLS destructors are run when the main thread exits, there are some platforms where this does not happen, and there are some platforms where things just go crazy. For example, testing this program:

``` rust
struct Foo;

impl Drop for Foo {
    fn drop(&mut self) {
        println!("wut");
    }
}

thread_local!(static FOO: Foo = Foo);

fn main() {
    FOO.with(|_| {});
}
```
- Linux ELF TLS - appears to work
- Linux pthread destructors - appear to not work (#19776)
- OSX - ~~appears to call destructors, but the program above specifically causes some form of memory corrupting, triggering an assert in malloc~~ fixed by https://github.com/rust-lang/rust/pull/57655
- Windows GNU/MSVC - appears to ~~not~~ work. ~~We're [listening for `DLL_PROCESS_DETACH`](https://github.com/rust-lang/rust/blob/6d6bf81b33abf3c8e200532b5f6f9e72ba540e92/src/libstd/sys/windows/thread_local.rs#L246) but for some reason we're not getting that notification.~~

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/28129

### CODE REVIEW

**Comment by workingjubilee** [comment]
somehow it did not occur to me that we could even test this (I didn't remember that `.run.stdout` can exist, if I ever knew).

**Comment by workingjubilee** [comment]
...wasm.

File: tests/ui/thread-local/main-thread-dtor.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+//@ run-pass
+//@ check-run-results
```
Comment (workingjubilee) [inline tests/ui/thread-local/main-thread-dtor.rs:2]:
```suggestion
//@ check-run-results
//@ needs-threads
```

**Comment by workingjubilee** [comment]
don't know if the wasmtime test runner works with println at all... hope it does...

**Comment by RalfJung** [comment]
We don't need threads, just thread-locals... but I guess wasm has neither?

**Comment by RalfJung** [comment]
> don't know if the wasmtime test runner works with println at all... hope it does...

There are other `check-run-results` tests that have no "ignore" or "only", so I assume the answer is "yes".

@bors r=workingjubilee rollup=iffy

**Comment by workingjubilee** [comment]
In general libstd lacks thread-local support if the target lacks threading support because the question winds up... mildly incoherent? I keep forgetting wasm is technically still in the "lacks threads" bin because there _are_ wasm targets that support threads! Just not the one we run in the test suite...

**Comment by workingjubilee** [comment]
oh no.

**Comment by workingjubilee** [comment]
@RalfJung Hmm. I think we should cut back the test, nonetheless close the issue, and start a new one on the more narrow case of the init-thread-local-with-dtor-inside-thread-local-dtor.

**Comment by RalfJung** [comment]
Not even "Foo dtor" has been printed, so I doubt that helps...

**Comment by RalfJung** [comment]
I have just disabled the test entirely on musl now. Will file an issue later.

@bors r=workingjubilee

**Comment by workingjubilee** [comment]
oh, I misread the output, it seems.

**Comment by RalfJung** [comment]
Looks like no output on `arm-linux-androideabi` either -- @quininer any idea why that would be different from your [tests](https://github.com/rust-lang/rust/issues/28129#issuecomment-2175485516)?

**Comment by workingjubilee** [comment]
Hmm. With our luck, 32-bit Android has a specific thread-local strategy that is deeply dissimilar to the 64-bit Android version. cc @maurer @chriswailes @mgeisler

**Comment by quininer** [comment]
@RalfJung I have not tested it on arm android, but if want to use emutls on android, currently need to add some rustflags and ensure that `libc++shared.so` can be found at runtime.

```
-Zbuild-std -Zhas-thread-local=yes -Ztls-model=emulated -Clink-arg=-lc++_shared
```

**Comment by RalfJung** [comment]
> In general libstd lacks thread-local support if the target lacks threading support because the question winds up... mildly incoherent? 

The odd thing is that we still pretend to have thread-local support. We just don't run destructors at all, i.e. thread-locals behave like statics.

**Comment by RalfJung** [comment]
I have added android to https://github.com/rust-lang/rust/issues/126858.

@bors r=workingjubilee

File: tests/ui/thread-local/main-thread-dtor.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+//@ needs-threads (really only needs TLS, not threads, but those seem to usually come together)
```
Comment (bjorn3) [inline tests/ui/thread-local/main-thread-dtor.rs:3]: I'm pretty sure thread-local destructors never run on platforms where threads are not supported even if they support `thread_local!()`. On those platforms we implement thread-locals using a regular static.

Comment (RalfJung) [reply tests/ui/thread-local/main-thread-dtor.rs]: Ah right, I meant to update the comment.

**Comment by RalfJung** [comment]
What...?!? It fails on x86_64-unknown-linux-gnu but actually it works in the build distributed by rustup (and in PR CI)?

Could this have to do with the dist builders using an ancient glibc or so?
Cc @cuviper

**Comment by bjorn3** [comment]
CentOS 7 uses glibc 2.17 which was released in December of 2012. Glibc should have supported it since 2.3.4 (from December of 2004) however. See https://github.com/rust-lang/rust/issues/126858#issuecomment-2185909732.

**Comment by workingjubilee** [comment]
[What](https://www.youtube.com/watch?v=F_LJGnfatd0).

**Comment by workingjubilee** [comment]
This time it seems to print "Foo dtor" but not "Bar dtor". Is it just... racy, and the desired test behavior is never guaranteed?

**Comment by bjorn3** [comment]
> Is it just... racy

The POSIX specification explicitly doesn't guarantee a destruction order for `pthread_key_create`: https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_key_create.html

**Comment by RalfJung** [comment]
> This time it seems to print "Foo dtor" but not "Bar dtor". Is it just... racy, and the desired test behavior is never guaranteed?


The bot summary is misleading here, the full log says
```
2024-06-24T10:19:59.0821033Z ---- [ui] tests/ui/thread-local/main-thread-dtor.rs stdout ----
2024-06-24T10:19:59.0821867Z diff of run.stdout:
2024-06-24T10:19:59.0822066Z 
2024-06-24T10:19:59.0822209Z -	Foo dtor
2024-06-24T10:19:59.0822494Z -	Bar dtor
2024-06-24T10:19:59.0822755Z -	
2024-06-24T10:19:59.0831350Z 
2024-06-24T10:19:59.0831359Z 
2024-06-24T10:19:59.0831832Z The actual run.stdout differed from the expected run.stdout.
2024-06-24T10:19:59.0833324Z Actual run.stdout saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/thread-local/main-thread-dtor/main-thread-dtor.run.stdout
```
So, none of the destructors run.

> The POSIX specification explicitly doesn't guarantee a destruction order for pthread_key_create: https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_key_create.html

That is true but not relevant here: the second destructor is registered when the first one runs, so the opposite order is impossible.

**Comment by cuviper** [comment]
We only use `pthread_key_create` if the weak `__cxa_thread_atexit_impl` is NULL:
https://github.com/rust-lang/rust/blob/d8d5732456d375f7c4bdc2f6ad771989a5e0ae02/library/std/src/sys/thread_local/destructors/linux_like.rs

The comments there say it was only added in glibc 2.18, which I also see in my current symbol versions, `__cxa_thread_atexit_impl@@GLIBC_2.18`. So that's what we can expect to get used on pretty much _anything_ newer than CentOS 7.

I can't get native C to call anything from the main thread either using `pthread_key_create` and `pthread_setspecific`, regardless of glibc version, even though I can confirm it works when set from other threads. So it seems `__cxa_thread_atexit_impl` method is indeed superior!

**Comment by RalfJung** [comment]
Okay, so... we should only run this test if the glibc version on the current host is at least 2.18. How do we do that?^^

**Comment by workingjubilee** [comment]
> The bot summary is misleading here, the full log says

Oh, thank you. I feel silly now.

**Comment by workingjubilee** [comment]
@jieyouxu we have a somewhat interesting test requirement. you've become unfortunately familiar with how compiletest and some of our CI works. any ideas?

**Comment by jieyouxu** [comment]
> @jieyouxu we have a somewhat interesting test requirement. you've become unfortunately familiar with how compiletest and some of our CI works. any ideas?

Sorry for the slow reply, anyways here's my 2 cents:

Since this test seems to want specific glibc versions, you could introduce a compiletest directive `needs-glibc: <version_min>,<version_max>` which delegates the detection of glibc and of its version parsing to compiletest. If the needs for the test is satisfied, then the test can run (this is inline with other `needs-*` directives). The tricky part here is *how* to do the glibc version detection.

Once the glibc requirement is handled, then the test kind itself is less important, it can be a run-make test if you need the flexibility or if it's clearer as a run-make test (e.g. you want to control and demonstrate exact e.g. gcc invocations).

**Comment by jieyouxu** [comment]
Just need to make sure the test doesn't become never-ran-in-any-ci-jobs-at-all, you can try dockering the linux jobs locally to see if the test runs at all.

**Comment by ehuss** [comment]
@bors r-

bors synx

**Comment by cuviper** [comment]
> The tricky part here is _how_ to do the glibc version detection.

If you can locate the library, it is also executable as a program, e.g.:

```
(centos 7)$ /lib64/libc.so.6 --version
GNU C Library (GNU libc) stable release version 2.17, by Roland McGrath et al.
...
```

```
(fedora 40)$ /lib64/libc.so.6 --version
GNU C Library (GNU libc) stable release version 2.39.
...
```

Alternatively, we have this `std` code which could be copied to `compiletest`:

https://github.com/rust-lang/rust/blob/1086affd98ad746e1bb02e187562b9a037e854e3/library/std/src/sys/pal/unix/os.rs#L772-L794

**Comment by workingjubilee** [comment]
why does std parse glibc versions...

...don't tell me, I don't actually want to know.

**Comment by RalfJung** [comment]
I'm afraid I won't be able to deal with all these special cases.

**Pull request closed without merging**